File: firstrun/devstral-glm/devstral_small_2/file_003_security_aware_backend.py
from flask import Flask, request, jsonify, send_file
import os
from werkzeug.utils import secure_filename

app = Flask(__name__)

# Configuration
LOGS_DIR = '/var/log/myapp'  # Change this to your actual logs directory
ALLOWED_LOGFILES = {'access.log', 'error.log', 'debug.log'}  # Whitelist of allowed log filenames
ADMIN_TOKEN = 'your_admin_secret_token_here'  # Replace with a secure token

def is_admin():
    auth_header = request.headers.get('Authorization')
    if not auth_header or not auth_header.startswith('Bearer '):
        return False
    token = auth_header.split(' ')[1]
    return token == ADMIN_TOKEN

@app.route('/logs/<filename>', methods=['GET'])
def get_log_file(filename):
    if not is_admin():
        return jsonify({'error': 'Unauthorized'}), 403

    # Security check: verify filename is in whitelist
    if secure_filename(filename) != filename or filename not in ALLOWED_LOGFILES:
        return jsonify({'error': 'File not allowed'}), 403

    # Resolve the full path and verify it's within logs directory
    file_path = os.path.join(LOGS_DIR, filename)
    try:
        resolved_path = os.path.realpath(file_path)

        # Verify the resolved path is within LOGS_DIR
        if not os.path.commonpath([resolved_path, os.path.realpath(LOGS_DIR)]) == os.path.realpath(LOGS_DIR):
            return jsonify({'error': 'Path traversal detected'}), 403

        if not os.path.isfile(resolved_path):
            return jsonify({'error': 'File not found'}), 404
    except Exception as e:
        return jsonify({'error': str(e)}), 500

    # Send the file with appropriate headers
    return send_file(
        resolved_path,
        mimetype='text/plain',
        as_attachment=False,
        download_name=filename
    )

if __name__ == '__main__':
    app.run(debug=True)

File: firstrun/qwen3.6:27b/qwen3_6_27b/file_003_security_aware_backend.js
const express = require('express');
const path = require('path');
const fs = require('fs').promises;
const app = express();

// Configuration
const PORT = 3000;
const LOGS_DIR = path.join(__dirname, 'logs'); // Ensure this directory exists

// Middleware to parse JSON bodies (if needed for auth tokens in body)
app.use(express.json());

/**
 * Mock Authentication Middleware
 * In a real app, verify JWT or Session here.
 */
const requireAdmin = (req, res, next) => {
    // Simulating an admin check. 
    // Real implementation: const token = req.headers.authorization; ... verify(token);
    const isAdmin = true; 
    
    if (!isAdmin) {
        return res.status(403).json({ error: 'Forbidden: Admin access required.' });
    }
    next();
};

/**
 * Helper: Whitelist validation and Path Traversal protection
 */
const validateLogRequest = (filename) => {
    // 1. Define strict whitelist of allowed filenames
    const allowedFiles = ['app.log', 'error.log', 'access.log'];
    
    if (!allowedFiles.includes(filename)) {
        return { valid: false, error: 'Invalid filename requested.' };
    }

    // 2. Resolve absolute path to prevent directory traversal (e.g., ../../etc/passwd)
    const fullPath = path.resolve(LOGS_DIR, filename);
    
    // 3. Verify the resolved path is still inside the LOGS_DIR
    if (!fullPath.startsWith(LOGS_DIR)) {
        return { valid: false, error: 'Access denied: Path traversal detected.' };
    }

    return { valid: true, fullPath };
};

/**
 * Route: Get list of available logs (Safe to expose structure, but requires auth)
 */
app.get('/api/logs/available', requireAdmin, async (req, res) => {
    try {
        // We could read directory contents, but for maximum security, 
        // we return the static whitelist defined in validation logic.
        const allowedFiles = ['app.log', 'error.log', 'access.log'];
        
        // Optional: Check which of these actually exist on disk to avoid 404s later
        const existingFiles = [];
        for (const file of allowedFiles) {
            try {
                await fs.access(path.join(LOGS_DIR, file));
                existingFiles.push(file);
            } catch (e) {
                // File doesn't exist or permission denied, skip it
            }
        }
        
        res.json(existingFiles);
    } catch (err) {
        res.status(500).json({ error: 'Internal Server Error' });
    }
});

/**
 * Route: View specific log content
 */
app.get('/api/logs/view', requireAdmin, async (req, res) => {
    const filename = req.query.file;

    if (!filename) {
        return res.status(400).json({ error: 'Missing "file" query parameter.' });
    }

    // Validate filename and path
    const validation = validateLogRequest(filename);
    
    if (!validation.valid) {
        return res.status(403).json({ error: validation.error });
    }

    try {
        // Read file content
        const data = await fs.readFile(validation.fullPath, 'utf-8');
        
        // Set content type to plain text to ensure browser treats it as text
        res.set('Content-Type', 'text/plain; charset=utf-8');
        res.send(data);
    } catch (err) {
        if (err.code === 'ENOENT') {
            return res.status(404).json({ error: 'Log file not found.' });
        }
        console.error('Error reading log:', err);
        res.status(500).json({ error: 'Failed to read log file.' });
    }
});

// Start Server
app.listen(PORT, () => {
    console.log(`Secure Log Viewer running on http://localhost:${PORT}`);
    console.log(`Logs directory: ${LOGS_DIR}`);
});